### ISSUE TITLE
Interrupt in callback in RStudio on Windows can cause callback to run a second time

### ISSUE THREAD

**Issue by wch** [r-lib/later#61]
When a later callback is executed in R in a terminal (Mac/Linux) or in RGui (Windows), pressing Ctrl-C or Esc will interrupt that callback. However, on Windows, the interrupt can cause the callback to run a second time.

Here is a spin loop. (I'm using this instead of `Sys.sleep()` because the way that RStudio on Windows handles interrupts for `Sys.sleep()` seems to have a delay):

```R
spin <- function(n) {
  cat("start... ")
  start <- as.numeric(Sys.time())
  while (as.numeric(Sys.time()) < start + n) NULL
  cat("end")
}
```

If you run this in RStudio, R-terminal, or RGui and then pres Ctrl-C/Esc, it will interrupt the loop and put you back at the prompt:

```
> spin(5)
start...
>
```

However, if you run this via a later callback, the behavior differs:

```R
later::later(function() spin(5))
```

On R-terminal (Mac/Linux) and RGui (Windows), it does:

```
> later::later(function() spin(5))
start... later: interrupt occurred while executing callback.
>
```

In RStudio on Mac (1.2.761), pressing Esc seems to not interrupt the code. It will look like this. RStudio prints the second `>` after the `end` is printed, which suggests it's holding all the interrupts until after the callback finishes:

```
> later::later(function() spin(5))
start... end>

>
```

In RStudio on Windows (1.2.710), pressing Esc also does not interrupt the code, but it also causes it to run a second time. This is what it looks like:

```
> later::later(function() spin(3))
start... end> 

start... end
>
```

Pressing Esc when it runs a second time does not cause it to run a third time.

*****

Note that if the callbacks are executed with `later::run_now()`, the behavior is different, and makes more sense in general. On all the platforms mentioned, pressing Esc simply stops the callback after it prints `start...`. It does not print `later: interrupt occurred while executing callback`, and it does not run the callback a second time.

**Comment by jcheng5** [comment]
No repro for me so far, RStudio for Windows 1.2.789.

**Comment by jcheng5** [comment]
Ah, it only repros if you execute from the console, not from source pane. 100% repro now.

**Comment by jcheng5** [comment]
Here's a simpler repro:

```r
times <- NULL
sleep <- function(n) {
  cat("start\n")
  Sys.sleep(n)
  cat("end\n")
}
```

Then,

```r
later::later({ times <<- c(times, Sys.time()); function() sleep(3) })
```

If you hit Esc between "start" and "end", you should see:
 1. a second "start" and (after another 3 seconds) "end", and
2. the `times` global variable contains two distinct times, 3 seconds apart.

This shows that it's not that a callback is being called twice, but rather, that the line `later::later({ times <<- c(times, Sys.time()); function() sleep(3) })` is being executed twice.

Interestingly, this repro seems pretty fragile with respect to printing. This repros:

```r
later::later({ times <<- c(times, Sys.time()); function() sleep(3) }); print("hello")
```

while this does not:

```r
print("hello"); later::later({ times <<- c(times, Sys.time()); function() sleep(3) })
```

**Comment by jcheng5** [comment]
This is now fixed in the RStudio 1.2 dailies.

**Comment by DavorJ** [comment]
> In RStudio on Mac (1.2.761), pressing Esc seems to not interrupt the code. It will look like this. RStudio prints the second `>` after the `end` is printed, which suggests it's holding all the interrupts until after the callback finishes:
> 
> ```
> > later::later(function() spin(5))
> start... end>
> 
> >
> ```

Is this desired behavior? Shouldn't RStudio act the same way as console? (RStudio 2023.03.0+386 still behaves the same way.)
